# Lab | Data Structuring and Combining Data

For Challenge 1, you can either use data which has already been cleaned (option 1) or clean it yourself (option 2).

## Challenge 1: Combining Data (Option 1)

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the following links:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2_clean.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3_clean.csv

Combine the data from the three dataframes into a single dataframe, named "customer_data", using appropriate merging, concatenating, and joining techniques.

Verify that the customer_data dataframe contains all the rows and columns from the three original dataframes.

In [34]:
import pandas as pd

data_1 = pd.read_csv('file1.csv')
data_2 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2_clean.csv')
data_3 = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3_clean.csv')

#file2 and file3 were not cleaned, therefore following cleaning measures are nessesary

def lowercase_column_names(df):
    # Convert column names to lowercase
    df.columns = [col.lower() for col in df.columns]
    
lowercase_column_names(data_2)
lowercase_column_names(data_3)

data_2 = data_2.rename(columns=lambda x: x.replace(' ', '_'))
data_3 = data_3.rename(columns=lambda x: x.replace(' ', '_'))

data_2 = data_2.drop(0, axis=0)
data_3 = data_3.drop(0, axis=0)

data_2 = data_2.rename(columns = {'st':'state'})
data_3 = data_3.rename(columns = {'st':'state'})

display(data_1.head())
display(data_2.head())
display(data_3.head())



customer       state gender             education  customer_lifetime_value  \
0  QZ44356     Arizona      F              Bachelor                697953.59   
1  AI49188      Nevada      F              Bachelor               1288743.17   
2  WW63253  California      M              Bachelor                764586.18   
3  GA49547  Washington      M  High School or Below                536307.65   
4  OC83172      Oregon      F              Bachelor                825629.78   

    income  monthly_premium_auto  number_of_open_complaints     policy_type  \
0      0.0                  94.0                        0.0   Personal Auto   
1  48767.0                 108.0                        0.0   Personal Auto   
2      0.0                 106.0                        0.0  Corporate Auto   
3  36357.0                  68.0                        0.0   Personal Auto   
4  62902.0                  69.0                        0.0   Personal Auto   

   vehicle_class  total_claim_amount  
0  Four-Door Car         1131.464935  
1   Two-Door Car          566.472247  
2            SUV          529.881344  
3  Four-Door Car           17.269323  
4   Two-Door Car          159.383042

customer       state gender             education customer_lifetime_value  \
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   
5  OC83172      Oregon      F              Bachelor              825629.78%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   
5  62902.0                  69.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323  
5   Two-Door Car          159.383042

customer       state gender             education customer_lifetime_value  \
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   
5  OC83172      Oregon      F              Bachelor              825629.78%   

    income  monthly_premium_auto number_of_open_complaints     policy_type  \
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   
5  62902.0                  69.0                    1/0/00   Personal Auto   

   vehicle_class  total_claim_amount  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323  
5   Two-Door Car          159.383042

In [25]:
def compare_columns(df1, df2, df3):
    if df1.columns.equals(df2.columns) and df2.columns.equals(df3.columns):
        if df1.columns.to_list() == df2.columns.to_list() == df3.columns.to_list():
            print("All three DataFrames have the same column names and order.")
        else:
            print("The column names are the same, but the order is different.")
    else:
        print("The column names are not the same in all three DataFrames.")
        
        # Print or write unmatched columns
        unmatched_df1 = df1.columns.difference(df2.columns.union(df3.columns))
        unmatched_df2 = df2.columns.difference(df1.columns.union(df3.columns))
        unmatched_df3 = df3.columns.difference(df1.columns.union(df2.columns))

        print("Unmatched columns in DataFrame 1:", unmatched_df1.to_list())
        print("Unmatched columns in DataFrame 2:", unmatched_df2.to_list())
        print("Unmatched columns in DataFrame 3:", unmatched_df3.to_list())

compare_columns(data_1, data_2, data_3)

All three DataFrames have the same column names and order.


In [35]:
def standardize_columns(df):
    # Replace values in 'gender' column
    df['gender'] = df['gender'].replace({'Femal': 'F', 'female': 'F', 'Male': 'M'})
    
    # Replace values in 'state' column
    df['state'] = df['state'].replace({'AZ': 'Arizona', 'Cali': 'California', 'WA': 'Washington'})
    
    # Replace values in 'education' column
    df['education'] = df['education'].replace({'Bachelors': 'Bachelor'})
    
    # Remove '%' from 'customer_lifetime_value' column
    df['customer_lifetime_value'] = df['customer_lifetime_value'].str.replace('%', '')
    
    # Replace values in 'vehicle_class' column
    df['vehicle_class'] = df['vehicle_class'].replace({'Sports Car': 'Luxury', 'Luxury SUV': 'Luxury', 'Luxury Car': 'Luxury'})
    
    return df

data_2 = standardize_columns(data_2)
data_3 = standardize_columns(data_3)

In [37]:
#transform to numeric
data_2['customer_lifetime_value'] = pd.to_numeric(data_2['customer_lifetime_value'], errors = 'coerce')
data_3['customer_lifetime_value'] = pd.to_numeric(data_3['customer_lifetime_value'], errors = 'coerce')

#fix the format
data_2['number_of_open_complaints'] = data_2['number_of_open_complaints'].apply(lambda x: int(str(x).split('/')[1]) if '/' in str(x) else x)
data_3['number_of_open_complaints'] = data_3['number_of_open_complaints'].apply(lambda x: int(str(x).split('/')[1]) if '/' in str(x) else x)

In [45]:
data_2 = data_2.dropna(how = 'all')
nulls_2 = data_2.isnull().sum()

data_3 = data_3.dropna(how = 'all')
nulls_3 = data_3.isnull().sum()

print(nulls_2)
print(nulls_3)


customer                       0
state                          0
gender                       121
education                      0
customer_lifetime_value        6
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64
customer                       0
state                          0
gender                       121
education                      0
customer_lifetime_value        6
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64


In [47]:
# Fill null values in customer_lifetime_value with the mean
mean_value_2 = data_2['customer_lifetime_value'].mean()
data_2['customer_lifetime_value'].fillna(mean_value_2, inplace=True)

mean_value_3 = data_3['customer_lifetime_value'].mean()
data_3['customer_lifetime_value'].fillna(mean_value_3, inplace=True)

nulls_2 = data_2.isnull().sum()
nulls_3 = data_3.isnull().sum()

print(nulls_2)
print(nulls_3)

customer                       0
state                          0
gender                       121
education                      0
customer_lifetime_value        0
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64
customer                       0
state                          0
gender                       121
education                      0
customer_lifetime_value        0
income                         0
monthly_premium_auto           0
number_of_open_complaints      0
policy_type                    0
vehicle_class                  0
total_claim_amount             0
dtype: int64


In [48]:
# Fill null values in 'gender' with 'U'
data_2['gender'].fillna('U', inplace=True)
data_3['gender'].fillna('U', inplace=True)

nulls_2 = data_2.isnull().sum()
nulls_3 = data_3.isnull().sum()

print(nulls_2)
print(nulls_3)

customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64
customer                     0
state                        0
gender                       0
education                    0
customer_lifetime_value      0
income                       0
monthly_premium_auto         0
number_of_open_complaints    0
policy_type                  0
vehicle_class                0
total_claim_amount           0
dtype: int64


In [53]:
# Check for duplicate rows
duplicate_rows_2 = data_2[data_2.duplicated()]
duplicate_rows_3 = data_3[data_3.duplicated()]

# Display the duplicate rows, if any
print("Duplicate Rows in data_2:")
display(duplicate_rows_2)

print("Duplicate Rows in data_3:")
display(duplicate_rows_3)

Duplicate Rows in data_2:


customer       state gender education  customer_lifetime_value   income  \
4008  GS98873     Arizona      F  Bachelor                323912.47  16061.0   
4009  CW49887  California      F    Master                462680.11  79487.0   
4010  MY31220  California      F   College                899704.02  54230.0   

      monthly_premium_auto number_of_open_complaints    policy_type  \
4008                  88.0                         0  Personal Auto   
4009                 114.0                         0   Special Auto   
4010                 112.0                         0  Personal Auto   

      vehicle_class  total_claim_amount  
4008  Four-Door Car               633.6  
4009            SUV               547.2  
4010   Two-Door Car               537.6

Duplicate Rows in data_3:


customer       state gender education  customer_lifetime_value   income  \
4008  GS98873     Arizona      F  Bachelor                323912.47  16061.0   
4009  CW49887  California      F    Master                462680.11  79487.0   
4010  MY31220  California      F   College                899704.02  54230.0   

      monthly_premium_auto number_of_open_complaints    policy_type  \
4008                  88.0                         0  Personal Auto   
4009                 114.0                         0   Special Auto   
4010                 112.0                         0  Personal Auto   

      vehicle_class  total_claim_amount  
4008  Four-Door Car               633.6  
4009            SUV               547.2  
4010   Two-Door Car               537.6

In [54]:
data_2 = data_2.drop_duplicates()
data_3 = data_3.drop_duplicates()

In [57]:
print(data_1.shape, data_2.shape, data_3.shape)

(1070, 11) (9133, 11) (9133, 11)


In [58]:
data_2.to_csv('file2.csv', index=False)
data_3.to_csv('file3.csv', index=False)

In [59]:
# Combine rows from all three DataFrames
customer_data = pd.concat([data_1, data_2, data_3], ignore_index=True)

# Print or display the combined DataFrame
print("Combined DataFrame:")
display(customer_data)

Combined DataFrame:


customer       state gender             education  \
0      QZ44356     Arizona      F              Bachelor   
1      AI49188      Nevada      F              Bachelor   
2      WW63253  California      M              Bachelor   
3      GA49547  Washington      M  High School or Below   
4      OC83172      Oregon      F              Bachelor   
...        ...         ...    ...                   ...   
19331  LA72316  California      M              Bachelor   
19332  PK87824  California      F               College   
19333  TD14365  California      M              Bachelor   
19334  UP19263  California      M               College   
19335  Y167826  California      M               College   

       customer_lifetime_value   income  monthly_premium_auto  \
0                 6.979536e+05      0.0                  94.0   
1                 1.288743e+06  48767.0                 108.0   
2                 7.645862e+05      0.0                 106.0   
3                 5.363077e+05  36357.0                  68.0   
4                 8.256298e+05  62902.0                  69.0   
...                        ...      ...                   ...   
19331             2.340599e+04  71941.0                  73.0   
19332             3.096511e+03  21604.0                  79.0   
19333             8.163890e+03      0.0                  85.0   
19334             7.524442e+03  21941.0                  96.0   
19335             2.611837e+03      0.0                  77.0   

      number_of_open_complaints     policy_type  vehicle_class  \
0                           0.0   Personal Auto  Four-Door Car   
1                           0.0   Personal Auto   Two-Door Car   
2                           0.0  Corporate Auto            SUV   
3                           0.0   Personal Auto  Four-Door Car   
4                           0.0   Personal Auto   Two-Door Car   
...                         ...             ...            ...   
19331                         0   Personal Auto  Four-Door Car   
19332                         0  Corporate Auto  Four-Door Car   
19333                         3  Corporate Auto  Four-Door Car   
19334                         0   Personal Auto  Four-Door Car   
19335                         0  Corporate Auto   Two-Door Car   

       total_claim_amount  
0             1131.464935  
1              566.472247  
2              529.881344  
3               17.269323  
4              159.383042  
...                   ...  
19331          198.234764  
19332          379.200000  
19333          790.784983  
19334          691.200000  
19335          369.600000  

[19336 rows x 11 columns]

In [60]:
# Check Rows
rows_check = len(data_1) + len(data_2) + len(data_3) == len(customer_data)

print("Rows Check:", rows_check)

Rows Check: True


# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [116]:
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv')
print(df.shape)

#delete unnamed column
df = df.drop(df.columns[0], axis=1)

cols = df.columns
column_names = pd.DataFrame(columns=cols)
print('Current column names:')
display(column_names.T)

print('Data types:')
display(df.dtypes)

non_numerical_columns = df.select_dtypes(include='object')
columns_to_exclude = ['effective_to_date', 'customer']
non_numerical_columns = non_numerical_columns.drop(columns=columns_to_exclude, errors='ignore')

df = df.drop(df.columns[0])
unique_values = {}
for column in non_numerical_columns.columns:
    unique_values[column] = non_numerical_columns[column].unique()

# Display unique values
for column, values in unique_values.items():
    print(f"\nUnique values for {column}:\n{values}")


(10910, 27)
Current column names:


Empty DataFrame
Columns: []
Index: [customer, state, customer_lifetime_value, response, coverage, education, effective_to_date, employmentstatus, gender, income, location_code, marital_status, monthly_premium_auto, months_since_last_claim, months_since_policy_inception, number_of_open_complaints, number_of_policies, policy_type, policy, renew_offer_type, sales_channel, total_claim_amount, vehicle_class, vehicle_size, vehicle_type, month]

Data types:


customer                          object
state                             object
customer_lifetime_value          float64
response                          object
coverage                          object
education                         object
effective_to_date                 object
employmentstatus                  object
gender                            object
income                             int64
location_code                     object
marital_status                    object
monthly_premium_auto               int64
months_since_last_claim          float64
months_since_policy_inception      int64
number_of_open_complaints        float64
number_of_policies                 int64
policy_type                       object
policy                            object
renew_offer_type                  object
sales_channel                     object
total_claim_amount               float64
vehicle_class                     object
vehicle_size                      object
vehicle_type    

KeyError: "['customer'] not found in axis"

In [117]:
display(df.head())

customer       state  customer_lifetime_value response  coverage education  \
0  DK49336     Arizona              4809.216960       No     Basic   College   
1  KX64629  California              2228.525238       No     Basic   College   
2  LZ68649  Washington             14947.917300       No     Basic  Bachelor   
3  XL78013      Oregon             22332.439460      Yes  Extended   College   
4  QA50777      Oregon              9025.067525       No   Premium  Bachelor   

  effective_to_date employmentstatus gender  income  ... number_of_policies  \
0        2011-02-18         Employed      M   48029  ...                  9   
1        2011-01-18       Unemployed      F       0  ...                  1   
2        2011-02-10         Employed      M   22139  ...                  2   
3        2011-01-11         Employed      M   49078  ...                  2   
4        2011-01-17    Medical Leave      F   23675  ...                  7   

      policy_type        policy  renew_offer_type  sales_channel  \
0  Corporate Auto  Corporate L3            Offer3          Agent   
1   Personal Auto   Personal L3            Offer4    Call Center   
2   Personal Auto   Personal L3            Offer3    Call Center   
3  Corporate Auto  Corporate L3            Offer2         Branch   
4   Personal Auto   Personal L2            Offer1         Branch   

   total_claim_amount  vehicle_class vehicle_size vehicle_type month  
0          292.800000  Four-Door Car      Medsize            A     2  
1          744.924331  Four-Door Car      Medsize            A     1  
2          480.000000            SUV      Medsize            A     2  
3          484.013411  Four-Door Car      Medsize            A     1  
4          707.925645  Four-Door Car      Medsize            A     1  

[5 rows x 26 columns]

In [68]:
#check NaNsb
nulls = df.isnull().sum()
print(nulls)

unnamed:_0                       0
customer                         0
state                            0
customer_lifetime_value          0
response                         0
coverage                         0
education                        0
effective_to_date                0
employmentstatus                 0
gender                           0
income                           0
location_code                    0
marital_status                   0
monthly_premium_auto             0
months_since_last_claim          0
months_since_policy_inception    0
number_of_open_complaints        0
number_of_policies               0
policy_type                      0
policy                           0
renew_offer_type                 0
sales_channel                    0
total_claim_amount               0
vehicle_class                    0
vehicle_size                     0
vehicle_type                     0
month                            0
dtype: int64


In [69]:
#check duplicates
duplicate_rows = df[df.duplicated()]
display(duplicate_rows)

Empty DataFrame
Columns: [unnamed:_0, customer, state, customer_lifetime_value, response, coverage, education, effective_to_date, employmentstatus, gender, income, location_code, marital_status, monthly_premium_auto, months_since_last_claim, months_since_policy_inception, number_of_open_complaints, number_of_policies, policy_type, policy, renew_offer_type, sales_channel, total_claim_amount, vehicle_class, vehicle_size, vehicle_type, month]
Index: []

[0 rows x 27 columns]

In [108]:
# Round up to the whole number
df['months_since_last_claim'] = df['months_since_last_claim'].apply(np.ceil)
df['number_of_open_complaints'] = df['number_of_open_complaints'].apply(np.ceil)

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

In [76]:
summary_table = df.pivot_table(values='monthly_premium_auto', index='sales_channel', aggfunc='sum').round(2)
summary_table['percentage'] = (summary_table['monthly_premium_auto'] / summary_table['monthly_premium_auto'].sum() * 100).round(2)
# Display the summary table
print("Summary Table:")
display(summary_table.reset_index())


Summary Table:


sales_channel  monthly_premium_auto  percentage
0         Agent                386335       38.00
1        Branch                280953       27.63
2   Call Center                197970       19.47
3           Web                151511       14.90

As we can see from the summary table, the most successful sales channel is 'Agent' with 38% of the total revenue. The least total revenue 151511 was calected through the 'Web'. The channels 'Branch' and 'Call Center' have 27.63% and 19.47% correspondingly.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [98]:
pivot_result = df.pivot_table(
    values='customer_lifetime_value',
    index=['education','gender'],
    aggfunc='mean'
).round(2)

pivot_result.columns = ['average_customer_lifetime_value']

display(pivot_result)

pivot_result = df.pivot_table(
    values='customer_lifetime_value',
    index=['gender','education'],
    aggfunc='mean'
).round(2)

pivot_result.columns = ['average_customer_lifetime_value']

display(pivot_result)

average_customer_lifetime_value
education            gender                                 
Bachelor             F                               7874.27
                     M                               7703.60
College              F                               7748.82
                     M                               8052.46
Doctor               F                               7328.51
                     M                               7415.33
High School or Below F                               8675.22
                     M                               8149.69
Master               F                               8157.05
                     M                               8168.83

average_customer_lifetime_value
gender education                                            
F      Bachelor                                      7874.27
       College                                       7748.82
       Doctor                                        7328.51
       High School or Below                          8675.22
       Master                                        8157.05
M      Bachelor                                      7703.60
       College                                       8052.46
       Doctor                                        7415.33
       High School or Below                          8149.69
       Master                                        8168.83

   Customers with 'High School or Below' education level have the highest average customer lifetime value across both gender    'Master' and 'Bachelor' education levels also contribute to high average valu       For each education level, the average customer lifetime value is different between ge        Female customers generally have higher average values compared to male customers across all education :
        While there are variations in average customer lifetime values across education levels, there is no clear linear pattern indicating that a specific education level consistently leads to higher or lowact:
        Customers with a 'Master' degree tend to have higher average customer lifetime values, particularly notable for females.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*

In [110]:
pivot_result = df.pivot_table(
    values='number_of_open_complaints',
    index='policy_type',
    columns='month',
    aggfunc='sum',
    fill_value=0
)

print("Pivot Result:")
display(pivot_result)

Pivot Result:


month              1     2
policy_type               
Corporate Auto   489   424
Personal Auto   1876  1599
Special Auto      92   102

In [118]:
df.to_csv('file4.csv', index=False)